### PR TITLE
add hard_negative_sampler.py for detection

### LINKED ISSUE #4300: Add HardNegativeSampler to suppress false positive, especially in detection
**Is your feature request related to a problem? Please describe.**
HardNegativeSampler is useful suppress false positive, especially in detection

**Describe the solution you'd like**
Add HardNegativeSampler

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### LINKED ISSUE #4405: integration error tests.test_retinanet
**Describe the bug**
torchscript compatibility in torch 1.9.1+cu102 torchvision 1.10.1
(no error with torch 1.8.1 and torch 1.10.2)

<details>

```
[2022-05-31T01:15:53.665Z] ======================================================================
[2022-05-31T01:15:53.665Z] ERROR: test_script_4 (tests.test_retinanet.TestRetinaNet)
[2022-05-31T01:15:53.665Z] ----------------------------------------------------------------------
[2022-05-31T01:15:53.665Z] Traceback (most recent call last):
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/parameterized/parameterized.py", line 533, in standalone_func
[2022-05-31T01:15:53.665Z]     return func(*(a + p.args), **p.kwargs)
[2022-05-31T01:15:53.665Z]   File "/home/jenkins/agent/workspace/Monai-pytorch-versions/tests/test_retinanet.py", line 161, in test_script
[2022-05-31T01:15:53.665Z]     test_script_save(net, data)
[2022-05-31T01:15:53.665Z]   File "/home/jenkins/agent/workspace/Monai-pytorch-versions/tests/utils.py", line 675, in test_script_save
[2022-05-31T01:15:53.665Z]     convert_to_torchscript(
[2022-05-31T01:15:53.665Z]   File "/home/jenkins/agent/workspace/Monai-pytorch-versions/monai/networks/utils.py", line 538, in convert_to_torchscript
[2022-05-31T01:15:53.665Z]     script_module = torch.jit.script(model, **kwargs)
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_script.py", line 1096, in script
[2022-05-31T01:15:53.665Z]     return torch.jit._recursive.create_script_module(
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_recursive.py", line 412, in create_script_module
[2022-05-31T01:15:53.665Z]     return create_script_module_impl(nn_module, concrete_type, stubs_fn)
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_recursive.py", line 474, in create_script_module_impl
[2022-05-31T01:15:53.665Z]     script_module = torch.jit.RecursiveScriptModule._construct(cpp_module, init_fn)
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_script.py", line 497, in _construct
[2022-05-31T01:15:53.665Z]     init_fn(script_module)
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_recursive.py", line 452, in init_fn
[2022-05-31T01:15:53.665Z]     scripted = create_script_module_impl(orig_value, sub_concrete_type, stubs_fn)
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_recursive.py", line 478, in create_script_module_impl
[2022-05-31T01:15:53.665Z]     create_methods_and_properties_from_stubs(concrete_type, method_stubs, property_stubs)
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_recursive.py", line 355, in create_methods_and_properties_from_stubs
[2022-05-31T01:15:53.665Z]     concrete_type._create_methods_and_properties(property_defs, property_rcbs, method_defs, method_rcbs, method_defaults)
[2022-05-31T01:15:53.665Z] RuntimeError: Can't redefine method: forward on class: __torch__.monai.apps.detection.networks.retinanet_network.RetinaNetClassificationHead (of Python compilation unit at: 0x3996730)
[2022-05-31T01:15:53.665Z] 
[2022-05-31T01:15:53.665Z] ======================================================================
[2022-05-31T01:15:53.665Z] ERROR: test_script_5 (tests.test_retinanet.TestRetinaNet)
[2022-05-31T01:15:53.665Z] ----------------------------------------------------------------------
[2022-05-31T01:15:53.665Z] Traceback (most recent call last):
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/parameterized/parameterized.py", line 533, in standalone_func
[2022-05-31T01:15:53.665Z]     return func(*(a + p.args), **p.kwargs)
[2022-05-31T01:15:53.665Z]   File "/home/jenkins/agent/workspace/Monai-pytorch-versions/tests/test_retinanet.py", line 161, in test_script
[2022-05-31T01:15:53.665Z]     test_script_save(net, data)
[2022-05-31T01:15:53.665Z]   File "/home/jenkins/agent/workspace/Monai-pytorch-versions/tests/utils.py", line 675, in test_script_save
[2022-05-31T01:15:53.665Z]     convert_to_torchscript(
[2022-05-31T01:15:53.665Z]   File "/home/jenkins/agent/workspace/Monai-pytorch-versions/monai/networks/utils.py", line 538, in convert_to_torchscript
[2022-05-31T01:15:53.665Z]     script_module = torch.jit.script(model, **kwargs)
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_script.py", line 1096, in script
[2022-05-31T01:15:53.665Z]     return torch.jit._recursive.create_script_module(
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_recursive.py", line 412, in create_script_module
[2022-05-31T01:15:53.665Z]     return create_script_module_impl(nn_module, concrete_type, stubs_fn)
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_recursive.py", line 474, in create_script_module_impl
[2022-05-31T01:15:53.665Z]     script_module = torch.jit.RecursiveScriptModule._construct(cpp_module, init_fn)
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_script.py", line 497, in _construct
[2022-05-31T01:15:53.665Z]     init_fn(script_module)
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_recursive.py", line 452, in init_fn
[2022-05-31T01:15:53.665Z]     scripted = create_script_module_impl(orig_value, sub_concrete_type, stubs_fn)
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_recursive.py", line 478, in create_script_module_impl
[2022-05-31T01:15:53.665Z]     create_methods_and_properties_from_stubs(concrete_type, method_stubs, property_stubs)
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_recursive.py", line 355, in create_methods_and_properties_from_stubs
[2022-05-31T01:15:53.665Z]     concrete_type._create_methods_and_properties(property_defs, property_rcbs, method_defs, method_rcbs, method_defaults)
[2022-05-31T01:15:53.665Z] RuntimeError: Can't redefine method: forward on class: __torch__.monai.apps.detection.networks.retinanet_network.RetinaNetClassificationHead (of Python compilation unit at: 0x3996730)
[2022-05-31T01:15:53.665Z] 
[2022-05-31T01:15:53.665Z] ======================================================================
[2022-05-31T01:15:53.665Z] ERROR: test_script_6 (tests.test_retinanet.TestRetinaNet)
[2022-05-31T01:15:53.665Z] ----------------------------------------------------------------------
[2022-05-31T01:15:53.665Z] Traceback (most recent call last):
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/parameterized/parameterized.py", line 533, in standalone_func
[2022-05-31T01:15:53.665Z]     return func(*(a + p.args), **p.kwargs)
[2022-05-31T01:15:53.665Z]   File "/home/jenkins/agent/workspace/Monai-pytorch-versions/tests/test_retinanet.py", line 161, in test_script
[2022-05-31T01:15:53.665Z]     test_script_save(net, data)
[2022-05-31T01:15:53.665Z]   File "/home/jenkins/agent/workspace/Monai-pytorch-versions/tests/utils.py", line 675, in test_script_save
[2022-05-31T01:15:53.665Z]     convert_to_torchscript(
[2022-05-31T01:15:53.665Z]   File "/home/jenkins/agent/workspace/Monai-pytorch-versions/monai/networks/utils.py", line 538, in convert_to_torchscript
[2022-05-31T01:15:53.665Z]     script_module = torch.jit.script(model, **kwargs)
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_script.py", line 1096, in script
[2022-05-31T01:15:53.665Z]     return torch.jit._recursive.create_script_module(
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_recursive.py", line 412, in create_script_module
[2022-05-31T01:15:53.665Z]     return create_script_module_impl(nn_module, concrete_type, stubs_fn)
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_recursive.py", line 474, in create_script_module_impl
[2022-05-31T01:15:53.665Z]     script_module = torch.jit.RecursiveScriptModule._construct(cpp_module, init_fn)
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_script.py", line 497, in _construct
[2022-05-31T01:15:53.665Z]     init_fn(script_module)
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_recursive.py", line 452, in init_fn
[2022-05-31T01:15:53.665Z]     scripted = create_script_module_impl(orig_value, sub_concrete_type, stubs_fn)
[2022-05-31T01:15:53.665Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_recursive.py", line 478, in create_script_module_impl
[2022-05-31T01:15:53.666Z]     create_methods_and_properties_from_stubs(concrete_type, method_stubs, property_stubs)
[2022-05-31T01:15:53.666Z]   File "/usr/local/lib/python3.8/dist-packages/torch/jit/_recursive.py", line 355, in create_methods_and_properties_from_stubs
[2022-05-31T01:15:53.666Z]     concrete_type._create_methods_and_properties(property_defs, property_rcbs, method_defs, method_rcbs, method_defaults)
[2022-05-31T01:15:53.666Z] RuntimeError: Can't redefine method: forward on class: __torch__.monai.apps.detection.networks.retinanet_network.RetinaNetClassificationHead (of Python compilation unit at: 0x3996730)
[2022-05-31T01:15:53.666Z] 
[2022-05-31T01:15:53.666Z] ----------------------------------------------------------------------
[2022-05-31T01:15:53.666Z] Ran 8788 tests in 1502.528s
[2022-05-31T01:15:53.666Z] 
```

</details>

### PR DESCRIPTION
**Pull request by Can-Zhao**
Signed-off-by: Can Zhao <canz@nvidia.com>

Fixes #4300.

Fixes #4405 

### Description
add hard_negative_sampler.py
make it a general sampler for classification tasks.
add unit test

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/apps/detection/hard_negative_sampler.py [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This script is modified from  https://github.com/MIC-DKFZ/nnDetection/blob/main/nndet/core/boxes/sampler.py
+
+Copyright 2020 Division of Medical Image Computing, German Cancer Research Center (DKFZ), Heidelberg, Germany
```
Comment (Nic-Ma) [inline monai/apps/detection/hard_negative_sampler.py:15]: @wyli Should we put the license information here?

Thanks.

File: monai/apps/detection/hard_negative_sampler.py [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+
+
+class HardNegativeSamplerMixin(ABC):
+    def __init__(self, pool_size: float = 10):
+        """
+        Create a pool from the highest scoring false positives and sample
+        defined number of negatives from it
+
+        Args:
+            pool_size (float): hard negatives are sampled from a pool of size:
```
Comment (Nic-Ma) [inline monai/apps/detection/hard_negative_sampler.py:75]: No need `(float)` in all the doc-strings of this PR.

Thanks.

**Comment by wyli** [comment]
Cc @mibaumgartner @dongyang0122 could you please have a look if you have time? this is part of the detection pipeline.

**Comment by wyli** [comment]
Cc @mibaumgartner @dongyang0122 could you please have a look if you have time? this is part of the detection pipeline.

File: monai/apps/detection/utils/hard_negative_sampler.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+                sampler = HardNegativeSampler(
+                    select_sample_size_per_image=6, positive_fraction=0.5, min_neg=1, pool_size=2
+                )
+                # two images with different number of samples
+                target_labels = [ torch.tensor([0,1]), torch.tensor([1,0,2,1])]
+                fg_probs = [ torch.rand(2), torch.rand(4)]
+                pos_idx_list, neg_idx_list = sampler.select_samples_img_list(target_labels, fg_probs)
+        """
+        pos_idx = []
+        neg_idx = []
```
Comment (mibaumgartner) [inline monai/apps/detection/utils/hard_negative_sampler.py:194]: nit: maybe asserting the same length could avoid some strange behaviour down the line

**Review by mibaumgartner — COMMENTED** [review]
Looks good to me 👍

**Review by wyli — APPROVED** [review]
this adds indices samplers based on the predictive scores.

**Pull request merged**